In [2]:
# requests from quotes.py not working so getting quotes here after downloading .html files from website manually
from bs4 import BeautifulSoup
import os.path
import pickle
from datetime import datetime
import calendar
import csv

from datetime import datetime
import pandas as pd

In [5]:
quotes = pd.read_csv('quotes.csv')

In [9]:
quotes.to_json('quotes.json')

In [10]:
# 2022
with open('365 Inspirational and Fascinating Quotes for 2022 - Inc. Australia.html') as f:
    soup = BeautifulSoup(f)

# h2_tags = soup.find_all(attrs={"class":"inc_article_body"})
h2_tags = soup.find_all('h2')
# len(h2_tags)

# Initialize lists to store dates and quotes
dates = []
quotes = []

# Iterate through each <h2> tag
for h2 in h2_tags:
    # Get the date from the <h2> tag
    date = h2.text.strip()
    # date = date.split(' (', 1)[0]
    # Sunday, January 01, 2024
    date_format = "%A, %B %d, %Y"

    parsed_date = datetime.strptime(date, date_format)
    dates.append(parsed_date)

    # parse date into format
    
    # Find the next <p> tag (which contains the quote)
    next_p = h2.find_next_sibling('p')
    
    # Get the quote from the <p> tag
    if next_p:
        quote = next_p.text.strip()# .encode('utf-8')
        quotes.append(quote)


with open('quotes_2022.csv', mode='w', newline='\n') as f:
    writer = csv.writer(f)
    writer.writerow(['Date', 'Quote'])  # write header
    writer.writerows(zip(dates, quotes))


In [19]:
# 2023
with open('365 Best Inspirational Quotes for 2023 - Inc. Australia.html') as f:
    soup = BeautifulSoup(f)

# h2_tags = soup.find_all(attrs={"class":"inc_article_body"})
h2_tags = soup.find_all('h2')
# len(h2_tags)

# Initialize lists to store dates and quotes
dates = []
quotes = []

# Iterate through each <h2> tag
for h2 in h2_tags:
    # Get the date from the <h2> tag
    date = h2.text.strip()
    # date = date.split(' (', 1)[0]
    # format: January 1, 2024 (Sunday)
    date_format = "%B %d, %Y (%A)"

    parsed_date = datetime.strptime(date, date_format)  # read the date based on the format
    parsed_date2 = parsed_date.strftime("%Y-%m-%d")   # save as date without the timestamp only the yyyy-mm-dd
    dates.append(parsed_date2)

    # parse date into format
    
    # Find the next <p> tag (which contains the quote)
    next_p = h2.find_next_sibling('p')
    
    # Get the quote from the <p> tag
    if next_p:
        quote = next_p.text.strip()# .encode('utf-8')
        quotes.append(quote)


with open('quotes_2023.csv', mode='w', newline='\n') as f:
    writer = csv.writer(f)
    writer.writerow(['Date', 'Quote'])  # write header
    writer.writerows(zip(dates, quotes))


In [21]:
# 2024
with open('366 Inspirational Quotes for 2024 - Inc. Australia.html') as f:
    soup = BeautifulSoup(f)

# h2_tags = soup.find_all(attrs={"class":"inc_article_body"})
h2_tags = soup.find_all('h2')
# len(h2_tags)

# Initialize lists to store dates and quotes
dates = []
quotes = []

# Iterate through each <h2> tag
for h2 in h2_tags:
    # Get the date from the <h2> tag
    date = h2.text.strip()
    # date = date.split(' (', 1)[0]
    date_format = "%A %B %d, %Y"

    parsed_date = datetime.strptime(date, date_format)  # read the date based on the format
    parsed_date2 = parsed_date.strftime("%Y-%m-%d")   # save as date without the timestamp only the yyyy-mm-dd
    dates.append(parsed_date2)

    # parse date into format
    
    # Find the next <p> tag (which contains the quote)
    next_p = h2.find_next_sibling('p')
    
    # Get the quote from the <p> tag
    if next_p:
        quote = next_p.text.strip()# .encode('utf-8')
        quotes.append(quote)


with open('quotes_2024.csv', mode='w', newline='\n') as f:
    writer = csv.writer(f)
    writer.writerow(['Date', 'Quote'])  # write header
    writer.writerows(zip(dates, quotes))


In [43]:

    

# 2020
with open('366 Top Inspirational Quotes and Motivational Quotes for Every Single Day in 2020 - Inc. Australia.html') as f:
    soup = BeautifulSoup(f)

# h2_tags = soup.find_all(attrs={"class":"inc_article_body"})
h2_tags = soup.find_all('h2')
# len(h2_tags)

# Initialize lists to store dates and quotes
dates = []
quotes = []
authors = []

# Iterate through each <h2> tag
for h2 in h2_tags:
    # Get the date from the <h2> tag
    date = h2.text.strip()
    date = date.split(' (', 1)[0]
    date_format = "%A, %B %d, %Y"

    parsed_date = datetime.strptime(date, date_format)  # read the date based on the format
    parsed_date2 = parsed_date.strftime("%Y-%m-%d")   # save as date without the timestamp only the yyyy-mm-dd
    dates.append(parsed_date2)

    # parse date into format
    
    # Find the next <p> tag (which contains the quote)
    next_p = h2.find_next_sibling('p')
    
    # Get the quote from the <p> tag
    if next_p:
        quote = next_p.text.strip()# .encode('utf-8')
        quotes.append(quote)
        


with open('quotes_2020.csv', mode='w', newline='\n') as f:
    writer = csv.writer(f)
    writer.writerow(['Date', 'Quote'])  # write header
    writer.writerows(zip(dates, quotes))


In [53]:
df = pd.read_csv('quotes.csv')
df.to_json('quotes.json', orient='records')

# Split author from quote text

In [3]:
df = pd.read_json('quotes.json')
df.head()

Date                                              Quote
0 2024-01-01  “Life becomes easier when you learn to accept ...
1 2024-01-02  “If you don’t fall, how are you going to know ...
2 2024-01-03  “You are in danger of living a life so comfort...
3 2024-01-04  “This life is yours. Take the power to choose ...
4 2024-01-05  “Take the opportunity to learn from your mista...

In [4]:
print(ord('—'))

8212


In [5]:
df['count 8212'] = df['Quote'].str.count(chr(8212))

In [6]:
# some have three...
df['count 8212'].describe()

count    1455.000000
mean        0.852921
std         0.389446
min         0.000000
25%         1.000000
50%         1.000000
75%         1.000000
max         3.000000
Name: count 8212, dtype: float64

In [28]:
df[df['count 8212'] == 1]['Quote'].str.split(chr(8212))[0]

['“Life becomes easier when you learn to accept the apology you never got.”',
 'R. Brault']

In [52]:
# get author from quote
df[['quote8212', 'author8212']] = df['Quote'].str.rsplit(chr(8212), n=1, expand=True).iloc[:, :2]

In [59]:
# These correctly output because of str.rsplit with n=1
df[df['count 8212'] > 1]

Date                                              Quote  \
771  2022-02-17  “You do have a story inside you; it lies artic...   
833  2022-04-20  “Far better is it to dare mighty things, to wi...   
836  2022-04-23  “Luck? I don’t know anything about luck. I’ve ...   
863  2022-05-20  “Love is our true destiny. We do not find the ...   
953  2022-08-18  “Even more, I had never meant to love him. One...   
975  2022-09-09  “Don’t take too much advice. Most people who h...   
1003 2022-10-07  “It’s important to be precise about words, bec...   
1009 2022-10-13  “I don’t look to jump over seven-foot bars — I...   
1032 2022-11-05  “Before you speak, run this through your head:...   
1105 2020-01-17  “Rarely are opportunities presented to you in ...   
1448 2020-12-25  “I hate that word — ‘lucky.’ It cheapens a lot...   

      count 8212                                          quote8212  \
771            2  “You do have a story inside you; it lies artic...   
833            3  “Far better is it to dare mighty things, to wi...   
836            2  “Luck? I don’t know anything about luck. I’ve ...   
863            2  “Love is our true destiny. We do not find the ...   
953            3  “Even more, I had never meant to love him. One...   
975            3  “Don’t take too much advice. Most people who h...   
1003           2  “It’s important to be precise about words, bec...   
1009           2  “I don’t look to jump over seven-foot bars — I...   
1032           2  “Before you speak, run this through your head:...   
1105           3  “Rarely are opportunities presented to you in ...   
1448           2  “I hate that word — ‘lucky.’ It cheapens a lot...   

                                author8212  
771                              Anne Rice  
833                     Theodore Roosevelt  
836                           Lucille Ball  
863                          Thomas Merton  
953                        Stephenie Meyer  
975   Ben Silbermann, founder of Pinterest  
1003                       Michael Nesmith  
1009                        Warren Buffett  
1032                         Beret Guidera  
1105                        Susan Wojcicki  
1448                        Peter Dinklage

In [60]:
# these have a different hyphen...
df[df['count 8212'] == 0]

Date                                              Quote  \
32   2024-02-02  “Man is free at the instant he wants to be.”–V...   
366  2023-01-01  “Life is a daring adventure or it is nothing a...   
369  2023-01-04  “Intuitively, you would think the more fairnes...   
370  2023-01-05  “Opportunities don’t happen. You create them.”...   
371  2023-01-06    “Nothing will work unless you do.”–Maya Angelou   
...         ...                                                ...   
722  2023-12-30  “It is amazing what you can accomplish if you ...   
723  2023-12-31  “The bad news is time flies. The good news is ...   
794  2022-03-12  “Before long, many people will telecommute fro...   
899  2022-06-25  “If you want one thing too much it’s likely to...   
1082 2022-12-25  “God never gives someone a gift they are not c...   

      count 8212                                          quote8212 author8212  
32             0  “Man is free at the instant he wants to be.”–V...       None  
366            0  “Life is a daring adventure or it is nothing a...       None  
369            0  “Intuitively, you would think the more fairnes...       None  
370            0  “Opportunities don’t happen. You create them.”...       None  
371            0    “Nothing will work unless you do.”–Maya Angelou       None  
...          ...                                                ...        ...  
722            0  “It is amazing what you can accomplish if you ...       None  
723            0  “The bad news is time flies. The good news is ...       None  
794            0  “Before long, many people will telecommute fro...       None  
899            0  “If you want one thing too much it’s likely to...       None  
1082           0  “God never gives someone a gift they are not c...       None  

[229 rows x 5 columns]

In [64]:
# ord("–") # 8211
df['count 8211'] = df['Quote'].str.count(chr(8211))

In [65]:
# get author from quote with different hyphen
df[['quote8211', 'author8211']] = df['Quote'].str.rsplit(chr(8211), n=1, expand=True).iloc[:, :2]
df

In [68]:
ord("—")

8212

In [76]:
# group them all together
df['quote by itself'] = df['quote8212']
df['author by itself'] = df['author8212'].fillna(df['author8211'])

In [77]:
df[df['quote by itself'].isna()]

Empty DataFrame
Columns: [Date, Quote, count 8212, quote8212, author8212, count 8211, quote8211, author8211, quote by itself, author by itself]
Index: []

In [78]:
df[df['author by itself'].isna()]

Empty DataFrame
Columns: [Date, Quote, count 8212, quote8212, author8212, count 8211, quote8211, author8211, quote by itself, author by itself]
Index: []

In [79]:
# doing visual inspection still comes up with quotes and author together?
df[['Date', 'quote by itself', 'author by itself']].to_csv('quotes.csv')

In [81]:
df.to_csv('quotes2.csv')

easier to do in excel/sheets by just copy and paste the 8211 quote in place of quote by itself for those 229 rows where author8212 is empty

In [62]:
# not an easy way to do it with the quotes
# df['count open quote'] = df['Quote'].str.count(chr(8220))
# df['count close quote'] = df['Quote'].str.count(chr(8221))
# df['count open quote'].sum(), df['count close quote'].sum(), df.shape[0]
# df = df.drop(columns=['count open quote', 'count close quote'])

In [ ]:
# fix what I can with code
# fix the rest by hand...

In [82]:
df = pd.read_csv('quotes3.csv')
df.head()

Unnamed: 0        Date                                              Quote  \
0           0  2024-01-01  “Life becomes easier when you learn to accept ...   
1           1  2024-01-02  “If you don’t fall, how are you going to know ...   
2           2  2024-01-03  “You are in danger of living a life so comfort...   
3           3  2024-01-04  “This life is yours. Take the power to choose ...   
4           4  2024-01-05  “Take the opportunity to learn from your mista...   

                                           quote8212          author8212  \
0  “Life becomes easier when you learn to accept ...           R. Brault   
1  “If you don’t fall, how are you going to know ...       Stephen Curry   
2  “You are in danger of living a life so comfort...       David Goggins   
3  “This life is yours. Take the power to choose ...  Susan Polis Schutz   
4  “Take the opportunity to learn from your mista...        Tony Robbins   

   count 8211                                          quote8211 author8211  \
0           0  “Life becomes easier when you learn to accept ...        NaN   
1           0  “If you don’t fall, how are you going to know ...        NaN   
2           0  “You are in danger of living a life so comfort...        NaN   
3           0  “This life is yours. Take the power to choose ...        NaN   
4           0  “Take the opportunity to learn from your mista...        NaN   

                                     quote by itself    author by itself  
0  “Life becomes easier when you learn to accept ...           R. Brault  
1  “If you don’t fall, how are you going to know ...       Stephen Curry  
2  “You are in danger of living a life so comfort...       David Goggins  
3  “This life is yours. Take the power to choose ...  Susan Polis Schutz  
4  “Take the opportunity to learn from your mista...        Tony Robbins

In [86]:
df['date'] = df['Date']
df['quote'] = df['quote by itself']
df['author'] = df['author by itself']

In [89]:
df[['date', 'quote', 'author']].to_json('quotes2.json', orient='records')